### DISCUSSION TITLE
Options to reduce home assistant and pfsense integration load

### DISCUSSION THREAD

**Discussion by markfrancisonly** [travisghansen/hass-pfsense#141] [Ideas]
I'm not using most of the entities the pfsense integration generates, and I want to reduce the impact to my firewall. The fact is, I cannot adequately manage pfsense from home assistant and certainly don't want to recreate the pfsense dashboard in home assistant. I can see the appeal in having a summary snapshot card with cpu or bandwidth data, and the last few firewall entries, but primarily I need a method to automate tasks only pfsense can perform. Using the integration, this comes at a cost.

Currently the pfsense integration creates 380 entities in my home, 295 disabled and *85 enabled - simply too many. 

How can the integration best be updated to reduce the number of  entities generated? Are more filter options are needed in config setup? device_trackers entities have a config setup filter toggle, which I greatly appreciate since I'm running a unifi controller integration that provides the network-based device trackers I need. 

Today, I'm only using the pfsense integration for WOL service... At peak usage, I've only used a single switch to toggle one firewall rule on/off. Due to this low usage and the huge number of entities, I uninstalled the integration today. Wow it felt great, but soon found that it was going to be a big headache to implement WOL broadcasting through my linux vlan/bonding/docker network infrastructure.... and pfsense can reliably send WOL packets.  

and how do I reduce the data refresh load on my firewall? The data refresh interval can be set to a very large number but then the entity states would always be stale. Eliminating the update interval with configuration could work, but this would require a service to reload (update) data. Are there any other uses for having a manual data update service?

Curious what are others' thoughts

**Comment by travisghansen** [comment]
What load do the unused entities introduce to your system? What is load introduced on the firewall vs when the integration is fully disabled?

**Reply by markfrancisonly** [reply]
Many examples of load. For instance, all entity states are pushed into the frontend using javascript, and there are many subscribers to state changes, even with the listener optimization magic the core ha dev team has implemented, I have several templates and scripts that filter or scan the entire entity table looking for matching states or attributes every second, etc. 

To give an idea of the problem, I've created a entity count table by domain:

My home assistant has 
| Domain              | Count |
|---------------------|-------|
| Alarm Control Panel | 2     |
| Automation          | 69    |
| Binary Sensor       | 202   |
| Button              | 57    |
| Camera              | 63    |
| Climate             | 2     |
| Cover               | 4     |
| Device Tracker      | 27    |
| Fan                 | 3     |
| Group               | 14    |
| Input Boolean       | 6     |
| Input Number        | 1     |
| Input Select        | 1     |
| Input Text          | 3     |
| Light               | 53    |
| Lock                | 1     |
| Media Player        | 58    |
| Number              | 15    |
| Person              | 6     |
| Proximity           | 1     |
| Remote              | 2     |
| Scene               | 20    |
| Script              | 71    |
| Sensor              | 677   |
| Sun                 | 1     |
| Switch              | 161   |
| Update              | 30    |
| Weather             | 2     |
| Zone                | 2     |
| Total               | 1554  |

Total dropped 85 when the pfsense integration was removed. 85 entities is not the whole picture, but every 85 unused entities count. 

With 69 automations and 71 scripts firing at random times, the issue is responsiveness on a server that is running several other containers. System doesn't bog, but I'm running a few low cost android wallpanels that would benefit from reduced load, etc, etc

**Reply by travisghansen** [reply]
Can the scripts not ignore disabled entities? That sounds odd to me that the unused entities are pushed to the ui anywhere besides the integration page where they can be enabled :(

**Reply by markfrancisonly** [reply]
The frontend has a hass object that get's populated at load and refreshed periodically via websocket for cards to respond to entity state changes, but these are only active entity states. 

Disabled entities do not load the system, but 'unavailable' entities are active entities.

also for me personally, I want to reduce the clutter and/or improve organization. I made the choice to move all entities into areas this weekend, to allow in-room behaviors. it was a chore even using csv files and python to make the updates. the effort left me with the feeling that fewer entities is better

**Reply by travisghansen** [reply]
A disabled entity should never be unavailable, so the number of unavailable entities should be extremely low (from this integration anyway).

Just so I understand you were trying to use some automatic tooling to set the rooms/areas vs using the UI? If you were using the UI  aren't the disabled entries completely out of site for doing such a chore?

**Reply by markfrancisonly** [reply]
generally, yes the disabled entities only clog up the user entity table from a admin management perceptive, which isn't really a problem except with 380 entities it is difficult to see what the integration is providing. the entities are not grouped in any way other than by name, and the names are long and difficult to read. 

As I mentioned, the pfsense integration is creating 85 pfsense active entities, and I'm no longer using any of them, but still like the idea of being able to toggle a rule on/off here in the future. Since I only use the WOL service to wake up my LG tv, and usually use my remote, I think I can live without that function for now

**Reply by travisghansen** [reply]
And disabling the 85 (or 84 of them) doesn’t solve your problem either?

To be clear I’m just trying to fully understand the root issue, not being critical of your desire/wish!

**Reply by markfrancisonly** [reply]
bumping my conclusion out of this deep thread...

**Comment by markfrancisonly** [comment]
just realized that I could select all to disable integration entities in ha. this and setting the scan interval to once a day, would be the easiest way to disable entity/refresh loading. after the scan interval any new firewall changes would create new entities and I'd need to repeat the process again. 

in this case, if I wanted to toggle a firewall rule, I would first need to enable the switch, or use a custom component to enable the switch (such as [ha-registry](https://github.com/amosyuen/ha-registry)) and wait the interval for new data on the result, or does flipping a switch force a data refresh for that entity or all entities?

still feel like there is a missed opportunity for a manual data refresh service

**Reply by travisghansen** [reply]
I really like the idea of a service to trigger the refresh. From a very high level the integration has 2 'coordinators' which house the data:

- the primary coordinator for everything except device_tacker data
- the device_tracker coordinator

Each of them is on a distinct scan interval. During each scan interval each runs several api requests and stores *all* the results in a python `dict` / structure. The important part is that the data is considered whole only after a series of api requests..and those api requests always happen jointly during the same 'refresh'.

It really just depends on the method/service being called about whether the integration triggers a refresh upon action or not. For example if you disable/enable a pfsense service/daemon it will trigger a refresh of the data immediately after the action has succeeded. Others may not...I'd have to audit each to know for sure :( If you depend solely on the integration to trigger those updates then the data should remain relatively fresh, however if you make changes in pfsense directly the changes would not be reflected until a refresh happens.

**Reply by markfrancisonly** [reply]
Is there a clever way to create a generic integration service that executes an arbitrary function on pfsense? If folks keep requesting new services, the integration could easily end up with 20 or many more services, which would be undesirable both for a home assistant instance and future code maintenance. 

a new generic service and manual data update service would do wonders to prevent indefinite scope expansion, minor requests would be answered without a new service declaration

**Reply by travisghansen** [reply]
Those already exist in `exec_command` and `exec_php`. Neither of them help tell the integration to refresh its data though.

**Comment by markfrancisonly** [comment]
for my use case, I would prefer to have only needed entities generated, and refresh only a targeted entity data, but I can see how someone else may want the integration to resync config and pull down the whole pfsense state in some cases. 

these days, my pfsense 2.6 dashboard takes a full 5 second delay to load in the browser, it's super annoying. while I haven't figured out the problem yet, presumably pfsense is waiting on one of the dashboard widgets to refresh data. pulling too much data at once is an anti-pattern

**Reply by travisghansen** [reply]
It’s a balancing act between more requests and more fine-grained data retrieval. For the vast majority of sensors I developed it in such a way as to gather all that data in a single api request (precisely for the reason of keeping the number of requests down).

Pulling too little data has its drawbacks as well.
